### ISSUE TITLE
[CN] Make type names consistent in spec language (eg `char` vs. `u8`)

### ISSUE THREAD

**Issue by septract** [rems-project/cerberus#442]
CN version: `git-74f5b374c [2024-07-25 21:16:44 +0100]`

CN uses both `u8` and `char` in the spec language, eg: 
```
char f1(char *p, char q); 
/*@ 
spec f(pointer p, u8 q); // <-- u8 here 
requires 
  take val = Owned<char>(p); // <-- char here 
ensures 
  return == (u8) val; // <-- u8 here 
@*/
```

For the sake of clarity, I think it would be better to use one consistently, rather than both. 

Similarly, `int` / `i32`:
```
int f2(int *p, int q); 
/*@ 
spec f2(pointer p, i32 q); 
requires 
  take val = Owned<int>(p); 
ensures 
  return == (i32) val; 
@*/
```

Alternatively, perhaps `char` / `u8` and similar types should be aliased? 

(I'm aware that the C spec does not in fact require that `char` is an `unsigned char`, nor that `char` is 8 bits wide)

**Comment by ZippeyKeys12** [comment]
> (I'm aware that the C spec does not in fact require that `char` is an `unsigned char`, nor that `char` is 8 bits wide)

You could change the CN types (`u8`, `i32`) to use the `stdint.h` names (`uint8_t`, `int32_t`).

**Comment by PeterSewell** [comment]
On Sat, 27 Jul 2024 at 01:25, Mike Dodds ***@***.***> wrote:

> CN version: git-74f5b374c [2024-07-25 21:16:44 +0100]
>
> CN uses both u8 and char in the spec language, eg:
>
> char f1(char *p, char q);
> /*@
> spec f(pointer p, u8 q); // <-- u8 here
> requires
>   take val = Owned<char>(p); // <-- char here
> ensures
>   return == (u8) val; // <-- u8 here
> @*/
>
> For the sake of clarity, I think it would be better to use one
> consistently, rather than both.
>
> Similarly, int / i32:
>
> int f2(int *p, int q);
> /*@
> spec f2(pointer p, i32 q);
> requires
>   take val = Owned<int>(p);
> ensures
>   return == (i32) val;
> @*/
>
> Alternatively, perhaps char / u8 and similar types should be aliased?
>
> (I'm aware that the C spec does not in fact require that char is an unsigned
> char, nor that char is 8 bits wide)
>
IIRC C char's now _are_ 8 bits wide.  Their signedness still might be
implementation-defined; not sure.

The CN u8 etc names are I think inspired both by the Linux kernel integer
type names and the Rust types.
Their CN semantics is different from the C stdint.h uint8_t etc., though,
so I don't think we should use those.

Certainly good for specs to be internally consistent, in any case.

**Comment by cp526** [comment]
The UPenn user feedback also had some related complaints.


Why does CN work like this? Originally CN expressed the semantics of C integers, and specifications about them, using unbounded integers (CN type `integer`). But, `integer` can't be used for specifying the layout of integers in memory, so necessarily type annotations for `Owned` had to be in terms of C types, not CN types. 

When moving to the current bitvector-based semantics we decided to make CN types distinguish signed and unsigned integers (rather than having a bitvector type without sign and having users explicitly specify whether they want the signed or unsigned variants of the relevant operations) to match C. 

So CN types are now expressive enough to capture all information present in C types (except for type qualifiers), which means there's more room for harmonising the types.

Option 1. We could replace the C-type annotations of `Owned` and `Block` with *CN* types, e.g. `Owned<u8>`. 

Option 2. We could allow users to write CN specifications in terms of C-types such as `char` and `int`, not just for resources but everywhere. CN would internally keep the existing language of types, but additionally expose in the frontend also C-types, as implementation-defined syntactic sugar for the corresponding internal types (e.g. the user's `char` is translated to `u8` internally).

The question is which one is more usable. 

Thomas S previously suggested option (2), with the argument that if C code is "parametric" in implementation choices such as the width of an `int` one may want to be equally parametric in CN specifications, perhaps even prove correctness for all implementation choices (e.g. for all reasonable widths of `int`).

Option (1) has the advantage that it makes the specifications clearer by avoiding the ambiguity in the widths and sign of C-types.


I'm leaning towards option (2). It reduces the gap between C and CN; where that's useful it still leaves open the option to write unambiguous specifications using CN types such as `u8` or `i32`; and it's not difficult to implement. Implementation-wise (2) is an extension of (1), the difference would only be frontend support for C-types in place of CN types, which would internally be converted into the existing CN types.


Either way, the problem has to be considered at the same time as other questions about CN's handling of integers:

- we moved CN to use bitvectors to be able to better handle bit manipulation in code and specifications; but writing correct specifications using bitvectors can be harder than with unbounded integers (e.g. avoiding over/underflows in specifications);

- ideally CN SMT performance can be improved enough that we could get predictable type checking performance with treating all C integers as bitvectors; if not, as Peter suggests, we may have to allow users to make fine-grained choices.

**Comment by bcpierce00** [comment]
I like both of these choices *way* better than the status quo!

Between the two, I feel like (2) is significantly better for users, since it avoids making people learn a whole separate language of types for the CN parts of the code.  If it's also easy to implement, it seems like a clear winner.

**Comment by septract** [comment]
I agree, option 2 seems better. 

It sounds like this would not invalidate any existing CN examples. Rather it would allow C types in positions that current require `u8` / `u16` or whatever. Is that right?

**Comment by cp526** [comment]
> I agree, option 2 seems better.
> 
> It sounds like this would not invalidate any existing CN examples. Rather it would allow C types in positions that current require `u8` / `u16` or whatever. Is that right?

Yes. What I'm imagining is effectively just an extension of the syntax of CN types: users can write C-types in place of CN types; internally these are desugared into the existing CN types. 

If we decided to go down this route -- though currently the interaction with the integer vs bitvector question is still unresolved -- we'd have to see whether the above desugaring scheme works or whether CN's core type system would have to be taught about those new C-types: e.g. if C-types are desugared into CN types, then (without extra work) error reports would be in terms of the existing CN types, not C-types.

**Comment by bcpierce00** [comment]
I was wondering about error reports.  Obviously, if these are in terms of
internal CN types instead of what the user actually wrote, there will be
lots of scope for confusion.  :-(

**Comment by septract** [comment]
Yeah, that would create confusion for sure. It sounds like we would probably want to eventually either map back to the syntax-level types, or actually implement those types in CN's core type system.

**Comment by cp526** [comment]
Perhaps we can have a mechanism for remembering the mapping from internal types back to user-written types. This is not a new problem, people have presumably figured out how to do this the right way for things like type abbreviations in other languages.

**Comment by cp526** [comment]
One more aspect to consider: the fact that C has an integer type called `bool` is a bit awkward here. CN needs a proper (non-integer) boolean type. If, as discussed above, we make the CN base types an extension of the C types – e.g. allowing users to write `char` and `int` in CN specifications (which would be nice UI-wise) – this clash of C booleans and CN booleans (however the latter should be named) becomes a bit more confusing.
